Posix: Add support for POSIX fstat

karthi012 opened the pull request:
Add support for POSIX fstat. fstat work on file descriptor, there is no way to retreive the filename from the zephyr filesystem, so adding filename variable fs_file struct to get the filename.

de-nordic commented on include/zephyr/fs/fs_interface.h:
```diff
@@ -61,6 +61,8 @@ struct fs_mount_t;
  */
 struct fs_file_t {
+	/** Name of the file */
+	char file_name[MAX_FILE_NAME];
```
That is significant increase in size of the fs_file_t structure, is it possible to offload that to POSIX layer, I am not sure that VFS layer is interested in taking the cost, at this point.

karthi012 replied:
@de-nordic  But it helps VFS layer too. There is no way to retreive the filename in current vfs implementation. I understand that it increase the fs_file_ struct, but it is from the application layer.

cfriedt replied:
Embedding the full filename like this is definitely not ideal.

There is an algorithm to piece together the full path given the inode, dirent info, and mount point.

https://stackoverflow.com/a/33987798

However, judging by [fs_stat()](https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/include/zephyr/fs/fs.h#L606) API and thr output struct, [fs_dirent](https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/include/zephyr/fs/fs.h#L117)

it really doesn't seem like it's sufficiently possible to fill in the majority of [struct stat](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_stat.h.html) using the `fs_stat()` call unless several additional items were added to the dentry structure (although that really is not where they belong).

I almost wonder if a new fs subsystem call will be required.

de-nordic replied:
> @de-nordic But it helps VFS layer too. There is no way to retreive the filename in current vfs implementation. I understand that it increase the fs_file_ struct, but it is from the application layer.

It does not seem there is an issue that requests the addition so far. 
As far as I can tell this is, now, required by POSIX, but there is not final client for that, no user case where addition/subsystem will finally request that kind of info.

The `fs_file_t` is very often allocated on stack or bss, this means that there will be huge addition to stack requirement (or bss)  for barely any use, as the code that does the definition of `fs_file_t` object really knows the file name and it will not even be interested to get it from the file object.
Basically: at the VFS API level users know names of files they tinker with or can store them in their own structures.

We have left all the `fs_file_t` object memory management to users/subsystems, because they know best when and how many objects they need, when the code gets into to the usage case.

POSIX has to pre-define, at compile time, how many (CONFIG_POSIX_MAX_FDS) entries the fd_table will be using and that is something you have tune for entire system; note that this is embedded system so you can not just tell anyone that now they have to get extra ram or do `ps` and find the thing that eats to many fds and kill it for other subsystem to work.

We could probably ifdef this with some Kconfig option, but note that, once you add this, all the `f_file_t` objects will have the array whether anything uses it or not; so if you have one use case for N objects the RAM tax for the usage case is `(N-1) * MAX_FILE_NAME`, for the VFS alone.

I think that rather than this change to `fs_file_t`, there should be RFC for the change explaining what are benefits, and what are costs so that the community could decide on the path forward. Or the `fdtable` should take the cost until VFS use case is found.

MaureenHelm commented:
@karthi012 please take a look at the failed CI checks

karthi012 committed:
lib: posix: Kconfig.fs: add kconfig for fstat

introduce POSIX_FILE_SYSTEM_FSTAT config to enable support
for POSIX fstat() functionality

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

karthi012 committed:
lib: posix: fs: introduce path variable

introduce path variable to store name of file for file system
operations based on the filename, there is no way to retreive
filename from Zephyr file system, to get stat info of a file,
filename is needed, so retreive of filename from fd is
offloaded to zephyr posix filesystem.

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

karthi012 committed:
lib: posix: fs: add filename to path variable

add name of file to path variable if POSIX_FILE_SYSTEM_FSTAT
is enabled

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

karthi012 committed:
lib: posix: fs: implement POSIX fstat()

implement POSIX fstat() functionality

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

karthi012 committed:
tests: posix: fs: add testuite for fstat()

add testsuite for POSIX fstat()

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

karthi012 commented:
@cfriedt  @de-nordic Please have look. I have offloaded filename to the POSIX implementation.

Please Ignore CI failures for now, i will take care of it.

ycsin commented:
> @cfriedt @de-nordic Please have look. I have offloaded filename to the POSIX implementation.
> 
> Please Ignore CI failures for now, i will take care of it.

meanwhile you can probably convert this to draft

karthi012 commented:
@cfriedt @de-nordic  Kind Ping.

cfriedt commented on lib/posix/options/fs.c:
```diff
@@ -157,6 +166,15 @@ static int fs_ioctl_vmeth(void *obj, unsigned int request, va_list args)
+		struct stat *buf = NULL;
+
+		buf = va_arg(args, struct stat *);
+		rc = stat(ptr->path, buf);
```
Based on my comments above, I think that the algorithm here is possibly backwards.

for `stat()`, given the absolute path of a file, we should

1. convert that to a tuple (mount point, node), where node is a pointer to an fs-specific handle, even just an offset into storage.
2. Use that tuple to call a file-system-specific method that returns data about the node.

For `fstat()`, we should
1. determine the mount point to generate the (mount point, node) tuple
2. Use that tuple to call the same file-system-specific method that returns data about the node.

cfriedt reviewed (changes requested):
See individual comments, but I think we have it kind of backwards; we shouldn't need an absolute path to look up file stats, just a tuple like (mount point or FS obj, and node).

For stat(), we should convert the absolute pathname to a tuple, and for fstat(), the node should be baked-into the fs-specific `filep` or `dirp` handle.

de-nordic commented on lib/posix/options/fs.c:
```diff
@@ -28,6 +29,9 @@ struct posix_fs_desc {
 	};
 	bool is_dir;
 	bool used;
+#if defined(CONFIG_POSIX_FILE_SYSTEM_FSTAT)
+	char path[MAX_FILE_NAME];
+#endif
```
> I'm not sure what the purpose of [`filep`](https://github.com/zephyrproject-rtos/zephyr/blob/7773fe8dd60e9f86760c994090304a0577a1ad2f/include/zephyr/fs/fs_interface.h#L78) is similarly the `dirp` field for directories, but to me that should probably represent a file-system-specific object that could maybe be used as a key.

You are correct regarding the what `dirp` and `filep` things are.
The filep and dirp are pointers to objects returned by FS drivers and VFS does not really understand them, but they have to be unique with respect to FS driver that serves them. There usually is no name to it or a way to associate it to name without storing that info in system. Even using f_open twice  on the same file puts you in risk of having two different filep objects, depending on design of underlying driver.

The VFS is constructed in a way that allows you to reuse one name buffer to open N files and operate on that N objects without remember the name anymore.

We are able to figure out mount points, because VFS stores them to figure out to which backend to talk them. Assuming I am  correct in the above statement we should have unique filep/dirp IDs within scope of mount point so, as you @cfriedt are suggesting that should be possible to match path to ID, but we still need a way to allocate memory for the record with name, somehow.

We could probably solve that with dynamic allocation where POSIX open would allocate common space for name. but that puts us again at MAX_FILE_NAME * number_of_ids cap, but now we are uncertain whether we or something else runs out of memory.

This is hard to solve.

cfriedt replied:
@de-nordic - is there a way to look up the parent `filep` or `dirp` for a given node?

If that were the case, and if each `filep` or `dirp` knew its own local name (which I assume would be encoded in the filesystem?) it should be possible to construct a full path with only the runtime overhead of traversing the FS tree, and the runtime allocation of a string of a given length for the full path.

Should be `O(n)` in terms of complexity, where n is the number of nodes in the path.

This would be really sweet.

cfriedt replied:
@de-nordic - is there a way to look up the parent `filep` or `dirp` for a given node?

If that were the case, and if each `filep` or `dirp` knew its own local name (which I assume would be encoded in the filesystem?) it should be possible to construct a full path with only the runtime overhead of traversing the FS tree, and the runtime allocation of a string of a given length for the full path.

Should be `O(n)` in terms of complexity, where n is the number of nodes in the path.

This would be really sweet.
